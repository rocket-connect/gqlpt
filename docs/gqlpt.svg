<svg width="100" height="100" viewBox="0 0 100 100" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="100" height="100" rx="16" fill="#0C0A09"/>
<path d="M70.9673 39.8168V11.5227H78.3874V33.5486H89V39.8168H70.9673Z" fill="url(#paint0_linear_676_164)"/>
<path d="M52.9181 34.2447C54.6149 34.2447 55.9379 33.4902 56.887 31.9812C57.8648 30.4722 58.3538 28.3247 58.3538 25.5388C58.3538 22.9561 57.8648 20.9682 56.887 19.5753C55.9379 18.1533 54.6149 17.4424 52.9181 17.4424C51.25 17.4424 49.927 18.1533 48.9491 19.5753C47.9713 20.9682 47.4824 22.9561 47.4824 25.5388C47.4824 28.3247 47.9713 30.4722 48.9491 31.9812C49.927 33.4902 51.25 34.2447 52.9181 34.2447ZM62.409 48C59.2741 48 56.6138 47.2455 54.428 45.7365C52.271 44.2275 50.646 42.2977 49.5531 39.9471C46.6196 39.2796 44.2756 37.698 42.5212 35.2024C40.7668 32.7067 39.8896 29.4855 39.8896 25.5388C39.8896 22.4337 40.4361 19.8075 41.529 17.66C42.6219 15.4835 44.1462 13.8294 46.1019 12.6976C48.0576 11.5659 50.3296 11 52.9181 11C55.5065 11 57.7785 11.5659 59.7342 12.6976C61.69 13.8294 63.2143 15.4835 64.3071 17.66C65.4 19.8075 65.9465 22.4337 65.9465 25.5388C65.9465 29.2243 65.17 32.2859 63.6169 34.7235C62.0926 37.1322 60.0075 38.7718 57.3615 39.6424C58.023 40.542 58.9002 41.1659 59.9931 41.5141C61.1147 41.8914 62.2076 42.08 63.2718 42.08C63.8757 42.08 64.4509 42.022 64.9974 41.9059C65.5438 41.8188 66.0328 41.7027 66.4642 41.5576L67.7584 46.9553C67.2119 47.2455 66.4785 47.4922 65.5582 47.6953C64.6379 47.8984 63.5881 48 62.409 48Z" fill="url(#paint1_linear_676_164)"/>
<path d="M26.1501 40.3388C23.5329 40.3388 21.1458 39.8165 18.9888 38.7718C16.8605 37.698 15.1636 36.0875 13.8982 33.94C12.6327 31.7926 12 29.1082 12 25.8871C12 23.5075 12.3739 21.4035 13.1217 19.5753C13.8694 17.718 14.8904 16.151 16.1846 14.8741C17.4788 13.5973 18.96 12.6396 20.6281 12.0012C22.325 11.3337 24.1081 11 25.9775 11C28.0483 11 29.8458 11.3773 31.3701 12.1318C32.8944 12.8573 34.1454 13.7278 35.1233 14.7435L31.2407 19.5318C30.5504 18.9224 29.817 18.429 29.0405 18.0518C28.2927 17.6455 27.358 17.4424 26.2364 17.4424C24.3382 17.4424 22.7564 18.1678 21.4909 19.6188C20.2254 21.0408 19.5927 23.0431 19.5927 25.6259C19.5927 28.2667 20.1823 30.3126 21.3615 31.7635C22.5694 33.1855 24.4245 33.8965 26.9266 33.8965C27.3292 33.8965 27.7175 33.8529 28.0914 33.7659C28.494 33.6788 28.8248 33.5482 29.0836 33.3741V29.4565H24.9421V23.3624H35.641V36.8565C34.6631 37.8141 33.3258 38.6412 31.6289 39.3377C29.9608 40.0051 28.1345 40.3388 26.1501 40.3388Z" fill="url(#paint2_linear_676_164)"/>
<path d="M12 89V52H25.2595C27.8965 52 30.3292 52.3795 32.5577 53.1385C34.7862 53.8974 36.569 55.1687 37.9061 56.9523C39.2803 58.7359 39.9674 61.1836 39.9674 64.2954C39.9674 67.2933 39.2803 69.76 37.9061 71.6954C36.569 73.5928 34.7862 74.9969 32.5577 75.9077C30.3664 76.8185 28.0079 77.2738 25.4823 77.2738H21.5825V89H12ZM21.5825 69.5323H25.0366C26.9308 69.5323 28.3236 69.0769 29.215 68.1662C30.1435 67.2174 30.6078 65.9272 30.6078 64.2954C30.6078 62.6256 30.1064 61.4492 29.1036 60.7662C28.1008 60.0831 26.6708 59.7415 24.8138 59.7415H21.5825V69.5323Z" fill="white"/>
<path d="M52.2593 89V60.1969H42.6769V52H71.4242V60.1969H61.8418V89H52.2593Z" fill="white"/>
<defs>
<linearGradient id="paint0_linear_676_164" x1="-240.701" y1="26.6535" x2="-132.746" y2="126.333" gradientUnits="userSpaceOnUse">
<stop stop-color="#FF7BCF"/>
<stop offset="1" stop-color="#F6009C"/>
</linearGradient>
<linearGradient id="paint1_linear_676_164" x1="-65.6461" y1="23.493" x2="14.7102" y2="86.2753" gradientUnits="userSpaceOnUse">
<stop stop-color="#FF7BCF"/>
<stop offset="1" stop-color="#F6009C"/>
</linearGradient>
<linearGradient id="paint2_linear_676_164" x1="11.787" y1="23.4931" x2="55.156" y2="35.7568" gradientUnits="userSpaceOnUse">
<stop stop-color="#FF7BCF"/>
<stop offset="1" stop-color="#F6009C"/>
</linearGradient>
</defs>
</svg>
