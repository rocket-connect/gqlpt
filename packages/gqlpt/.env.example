TEST_API_KEY="CHATGPT-API_KEY"