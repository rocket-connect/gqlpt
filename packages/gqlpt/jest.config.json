{
  "preset": "ts-jest/presets/default-esm",
  "extensionsToTreatAsEsm": [".ts"],
  "verbose": true,
  "testEnvironment": "node",
  "transform": {
    "^.+\\.(ts|tsx)?$": ["ts-jest", { "useESM": true }]
  },
  "testPathIgnorePatterns": ["./dist"],
  "testTimeout": 60000
}
