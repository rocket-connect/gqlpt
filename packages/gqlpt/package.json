{
  "name": "gqlpt",
  "version": "0.0.0-alpha.4",
  "description": "Leverage AI to generate GraphQL queries from plain text.",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "files": [
    "build"
  ],
  "scripts": {
    "test": "NODE_ENV=test jest",
    "build": "tsc"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rocket-connect/gqlpt.git"
  },
  "keywords": [],
  "author": "Rocket Connect - www.rocketconnect.co.uk",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/rocket-connect/gqlpt/issues"
  },
  "homepage": "https://github.com/rocket-connect/gqlpt#readme",
  "peerDependencies": {
    "graphql": "^16.0.0 || ^17.0.0"
  },
  "dependencies": {
    "openai": "4.11.0"
  }
}
