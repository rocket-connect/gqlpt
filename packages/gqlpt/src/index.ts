import { Adapter } from "@gqlpt/adapter-base";
import { introspection, postGeneratedQuery } from "@gqlpt/utils";

import { buildClientSchema, parse, print, printSchema } from "graphql";

import { DefaultTypeMap, GeneratedTypeMap } from "./types";

export interface GQLPTClientOptions {
  url?: string;
  headers?: Record<string, string>;
  typeDefs?: string;
  adapter: Adapter;
}

type MergedTypeMap = GeneratedTypeMap & DefaultTypeMap;

export class GQLPTClient<T extends MergedTypeMap = MergedTypeMap> {
  private options: GQLPTClientOptions;

  constructor(options: GQLPTClientOptions) {
    this.options = options;

    if (!options.adapter) {
      throw new Error("Missing adapter");
    }

    if (!options.typeDefs && !options.url) {
      throw new Error("Missing typeDefs or url");
    }

    if (options.typeDefs) {
      try {
        parse(options.typeDefs);
      } catch (error) {
        throw new Error(`Cannot parse typeDefs ${error}`);
      }
    }
  }

  public getTypeDefs() {
    return this.options.typeDefs;
  }

  async connect() {
    await this.options.adapter.connect();

    if (this.options.url) {
      const response = await introspection({
        url: this.options.url,
        headers: this.options.headers,
      });

      const schema = buildClientSchema(response.data);

      this.options.typeDefs = printSchema(schema);
    }
  }

  async generateQueryAndVariables(plainText: string): Promise<{
    query: string;
    variables?: Record<string, unknown>;
  }> {
    if (!this.options.typeDefs) {
      throw new Error("Missing typeDefs");
    }

    const query = `
      With this graphql schema: '${this.options.typeDefs}',
      and this question: '${plainText}',  
      generate a JSON object, where 'query' is a GraphQL query
      and 'variables' is a object containing all the variables. 
      Make sure you declare all the graphql variables in the query.
      Only traverse nested objects if they are explicitly requested.
      Dont add any more text or formating to your response, I will JSON parse the text.
    `;

    const response = await this.options.adapter.sendText(query);

    const result = JSON.parse((response || "").replace(/`/g, "")) as {
      query: string;
      variables?: Record<string, unknown>;
    };
    const queryAst = parse(result.query, { noLocation: true });
    const printedQuery = print(queryAst);

    return {
      query: printedQuery,
      variables: result.variables,
    };
  }

  async generateAndSend<Q extends string>(
    plainText: Q,
    {
      urlOverride,
      headersOverride,
    }: {
      urlOverride?: string;
      headersOverride?: Record<string, string>;
    } = {},
  ): Promise<Q extends keyof T ? T[Q] : any> {
    if (!this.options.url && !urlOverride) {
      throw new Error("Missing url");
    }

    const { query, variables } =
      await this.generateQueryAndVariables(plainText);

    const response = await postGeneratedQuery({
      query,
      variables,
      url: (urlOverride || this.options.url) as string,
      headers: headersOverride || this.options.headers,
    });

    return response as Q extends keyof T ? T[Q] : any;
  }

  async generateTypeMap(
    plainTextStrings: string[],
  ): Promise<Record<string, string>> {
    if (!this.options.typeDefs) {
      throw new Error("Missing typeDefs");
    }

    const query = `
      Given the following GraphQL schema:
      
      ${this.options.typeDefs}

      And these plain text queries:
      ${plainTextStrings.map((q, i) => `${i + 1}. ${q}`).join("\n")}

      Generate a JSON object where:
      - Each key is the original plain text query
      - Each value is a TypeScript type definition that represents the expected structure of the query result

      Rules for generating type definitions:
      1. Use TypeScript syntax
      2. Nest types inline (don't use separate interface declarations)
      3. Use specific types (string, number, boolean) where appropriate
      4. Use arrays ([]); when a field can return multiple items
      5. Make properties optional (?) if they might not always be present
      6. Use 'any' only as a last resort for unknown types

      Example output format:
      {
        "find users with name and email": "{ users: { name: string; email: string; }[] }",
        "get user by id": "{ user: { id: string; name: string; email?: string; } | null }"
      }

      Provide only the JSON object in your response, with no additional text or formatting.
    `;

    const response = await this.options.adapter.sendText(query);

    const result = JSON.parse((response || "").replace(/`/g, "")) as Record<
      string,
      string
    >;

    return result;
  }
}
