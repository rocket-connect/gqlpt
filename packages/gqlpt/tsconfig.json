{
  "compilerOptions": {
    "lib": ["ES2022"],
    "target": "ES2022",
    "module": "ES2022",
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "baseUrl": "./",
    "outDir": "build",
    "declarationMap": true,
    "declaration": true,
    "rootDir": "./src",
    "types": ["jest", "node", "@types/jest"],
    "sourceMap": true,
    "composite": true
  },
  "ts-node": {
    "esm": true,
    "experimentalSpecifierResolution": "node"
  },
  "include": ["./src/**/*"]
}
