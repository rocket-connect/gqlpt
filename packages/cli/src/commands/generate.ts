import { AdapterAnthropic } from "@gqlpt/adapter-anthropic";
import { Adapter } from "@gqlpt/adapter-base";
import { AdapterOpenAI } from "@gqlpt/adapter-openai";

import { Command } from "commander";
import fs from "fs/promises";
import { GQLPTClient } from "gqlpt";
import path from "path";

import { generateTypes } from "../generator";
import { parseFiles } from "../parser";

export const generate = new Command("generate");

generate
  .description("Generate type definitions for GQLPT")
  .argument("<source>", "Source directory to scan for GQLPT usage")
  .option(
    "-a, --adapter <adapter>",
    "The type of adapter to use either OpenAI or Anthropic",
    "openai",
  )
  .option("-o, --output <path>", "Output path for generated types")
  .option("-t --typeDefs <typeDefs>", "Path to type definitions file")
  .option("-u --url <url>", "GraphQL server URL")
  .option("-h --headers <headers>", "Headers to send to the server")
  .option("-r --raw", "Raw type definitions to stdout")
  .action(async (source, options) => {
    let adapter: Adapter;

    switch (options.adapter) {
      case "openai":
        if (!process.env.OPENAI_API_KEY) {
          throw new Error("process.env.OPENAI_API_KEY is required");
        }
        adapter = new AdapterOpenAI({
          apiKey: process.env.OPENAI_API_KEY,
        });
        break;
      case "anthropic":
        if (!process.env.ANTHROPIC_API_KEY) {
          throw new Error("process.env.ANTHROPIC_API_KEY is required");
        }
        adapter = new AdapterAnthropic({
          apiKey: process.env.ANTHROPIC_API_KEY,
        });
        break;
      default:
        throw new Error("Invalid adapter");
    }

    let outputPath: string;
    if (options.output) {
      outputPath = path.resolve(process.cwd(), options.output);
    } else {
      outputPath = path.resolve(
        process.cwd(),
        "node_modules/gqlpt/build/types.d.ts",
      );
    }

    let typeDefs: string = "";
    if (options.typeDefs) {
      typeDefs = await fs.readFile(options.typeDefs, "utf-8");
    }

    let headers: undefined | Record<string, string> = undefined;
    if (options.headers) {
      headers = JSON.parse(options.headers);
    }

    const client = new GQLPTClient({
      adapter,
      typeDefs,
      url: options.url,
      headers,
    });
    await client.connect();

    const srcDir = path.resolve(process.cwd(), source);

    const files = await getTypeScriptFiles(srcDir);

    const queries = await parseFiles(files);

    const typesContent = await generateTypes({
      queries: queries.map((x) => x.query),
      client,
    });

    if (options.raw) {
      process.stdout.write(typesContent);
    } else {
      await fs.writeFile(outputPath, typesContent);
      console.log("Type generation complete!");
    }
  });

async function getTypeScriptFiles(dir: string): Promise<string[]> {
  const entries = await fs.readdir(dir, { withFileTypes: true });
  const files = await Promise.all(
    entries.map((entry) => {
      const res = path.resolve(dir, entry.name);
      return entry.isDirectory() ? getTypeScriptFiles(res) : res;
    }),
  );
  return files
    .flat()
    .filter((file) => file.endsWith(".ts") || file.endsWith(".tsx"));
}
