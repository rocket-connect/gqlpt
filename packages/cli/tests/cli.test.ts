import dotenv from "dotenv";
import { GQLPTClient } from "gqlpt";
import path from "path";

import { generate } from "../src/commands/generate";

dotenv.config();

const OPENAI_API_KEY = process.env.TEST_OPENAI_API_KEY as string;

describe("generate", () => {
  const fixturePath = path.resolve(__dirname, "fixture/");
  const schemaPath = path.resolve(fixturePath, "schema.gql");

  test("should with generate types with openai adapter", async () => {
    const stdMock = jest.spyOn(process.stdout, "write").mockImplementation();

    const connectionSpy = jest
      .spyOn(GQLPTClient.prototype, "connect")
      .mockImplementation(async () => {});

    try {
      await generate.parseAsync([
        "generate",
        "--source",
        fixturePath,
        "--typeDefs",
        schemaPath,
        "--adapter",
        "openai",
        "--key",
        OPENAI_API_KEY,
        "--raw",
      ]);

      expect(connectionSpy).toHaveBeenCalledTimes(1);
      const content = stdMock.mock.calls[0][0];
      expect(content).toMatchSnapshot("");
    } finally {
      stdMock.mockRestore();
      connectionSpy.mockRestore();
    }
  });
});
