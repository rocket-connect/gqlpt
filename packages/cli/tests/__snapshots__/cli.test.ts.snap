// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate should with generate types with openai adapter 1`] = `
"
// This file is auto-generated. Do not edit manually.
// This will be populated by the CLI if used
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface GeneratedTypeMap {
  "Find users by id 1": { errors?: any[]; data: { user?: { id: string; name: string; } } };
  "find posts by user id 2": { errors?: any[]; data: { posts: [{ id: string; title: string; body: string; }] } };
}

// Default type map for when CLI is not used or for fallback
export type DefaultTypeMap = Record<string, any>;
//# sourceMappingURL=types.d.ts.map
"
`;
